### DISCUSSION TITLE
note-on and note-off

### DISCUSSION THREAD

**Discussion by BilldarBagdar** [markusschloesser/MackieC4_P3#84] [Q&A]
>  it will never be a note off message   
> MS: not true! Live clearly shows 2 msg, one when pressing, another on when releasing.

I can double check empirically, but I meant "never a note off" in a very technical sense.  The C4 sends true midi "note off" messages when you let off a button click, in the 0x80 hex range.  But in this[ receive_midi method](https://github.com/markusschloesser/MackieC4_P3/blob/master/wip/MackieC4/MackieC4.py#194), I never saw 0x80 messages logged.

There is no logical or audible difference between a "note on" message (0x90 range) at velocity 0 and a true "note off" message, so a lot of what I think I meant was Live translates "note off" from the C4 to "note on" at velocity 0.

I think the rest is related to the "modifier" buttons, so switching to that issue...

**Comment by BilldarBagdar** [comment]
Good thing I double checked...  :)  I uncommented some log messages and the "note off" messages from the C4  from these 4 modifier buttons are coming through loud and clear.  I still remember very sincerely writing that comment though...

```
(MackieC4) noteON<True> noteOFF<False> cc<False> received
(MackieC4) note<13> velo<127> received
(MackieC4) noteON<False> noteOFF<True> cc<False> received
(MackieC4) note<13> velo<0> received
(MackieC4) noteON<True> noteOFF<False> cc<False> received
(MackieC4) note<14> velo<127> received
(MackieC4) noteON<False> noteOFF<True> cc<False> received
(MackieC4) note<14> velo<0> received
(MackieC4) noteON<True> noteOFF<False> cc<False> received
(MackieC4) note<15> velo<127> received
(MackieC4) noteON<False> noteOFF<True> cc<False> received
(MackieC4) note<15> velo<0> received
(MackieC4) noteON<True> noteOFF<False> cc<False> received
(MackieC4) note<16> velo<127> received
(MackieC4) noteON<False> noteOFF<True> cc<False> received
(MackieC4) note<16> velo<0> received
```

**Reply by BilldarBagdar** [reply]
also looks like that comment had it completely backwards - it is the C4 that sends "note on" and velocity "0" instead of "note off" when a button is released; and Live translates the event to "note off" and velocity "0".

**Reply by markusschloesser** [reply]
ah, I misunderstood, I though you meant "C4 doesn't send (note off/velocity 0) at all :-)

sorry, been extremely busy with work the last 2 weeks, and this is gonna continue for at least another 2. thats why I've been quite quiet
